cns/restserver/util.go

```diff
@@ -235,29 +236,52 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
 		}
 	}
 
-	// Add the newIpConfigs, ignore if ip state is already in the map
-	service.addIPConfigStateUntransacted(req.NetworkContainerid, newIPConfigs)
+	// Add the newIpConfigs for 3 different conditions
+	// 1. If NC version from NMAgent is larger or same as new NC version, add new IP as available.
+	// 2. If NC version from NMAgent is larger or same as existing NC version, set existing pending programming IP as available and add new IP as pending programming.
+	// 3. If NC version from NMAgent is smaller than existing NC version, add new IP as pending programming.
+	var newNCVersion int
+	var existingNCVersion int
+	newNCVersion, _ = strconv.Atoi(req.Version)
+	existingNCVersion, _ = strconv.Atoi(existingNCVersionInString)
+	// Wait for Ashvin's thread update which querying NMAgent to get latest NC version and save it locally.
+	nmAgentNCVersion := getNCVersionFromNMAgent(req.NetworkContainerid)
+	if nmAgentNCVersion >= newNCVersion {
+		service.addIPConfigStateUntransacted(cns.Available, cns.Available, req.NetworkContainerid, newIPConfigs)
+	} else {
+		if nmAgentNCVersion >= existingNCVersion {
+			service.addIPConfigStateUntransacted(cns.Available, cns.PendingProgramming, req.NetworkContainerid, newIPConfigs)
+		} else {
+			service.addIPConfigStateUntransacted(cns.PendingProgramming, cns.PendingProgramming, req.NetworkContainerid, newIPConfigs)
+		}
+	}
+
+	existingNCStatus, ok := service.state.ContainerStatus[req.NetworkContainerid]
+	if ok {
+		existingNCStatus.HostVersion = strconv.Itoa(nmAgentNCVersion)
+	}
 
 	return 0, ""
 }
 
 // addIPConfigStateUntransacted adds the IPConfigs to the PodIpConfigState map with Available state
 // If the IP is already added then it will be an idempotent call. Also note, caller will
 // acquire/release the service lock.
-func (service *HTTPRestService) addIPConfigStateUntransacted(ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
+func (service *HTTPRestService) addIPConfigStateUntransacted(updatedExistingIPCNSStatus, newIPCNSStatus, ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
 	// add ipconfigs to state
 	for ipId, ipconfig := range ipconfigs {
-		// if this IPConfig already exists in the map, then ignore as this is an idempotent state
-		if _, exists := service.PodIPConfigState[ipId]; exists {
+		if existingPodIpConfigStatus, exists := service.PodIPConfigState[ipId]; exists {
+			if existingPodIpConfigStatus.State == cns.PendingProgramming {
+				existingPodIpConfigStatus.State = updatedExistingIPCNSStatus
```

Comment:
Can you add a log in case existing state is getting updated?

Reply:
I can.